File: backend/src/main/java/com/vacanza/backend/entity/UserPreferences.java
package com.vacanza.backend.entity;

import com.vacanza.backend.entity.enums.*;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UuidGenerator;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

/**
 * Explicit user preferences for AI personalization.
 * One-to-one with User (same pattern as UserInfo).
 */
@Entity
@Table(name = "user_preferences", uniqueConstraints = @UniqueConstraint(name = "uk_preferences_user", columnNames = "user_id"))
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class UserPreferences {

    @Id
    @GeneratedValue
    @UuidGenerator
    @Column(name = "preferences_id", nullable = false, updatable = false)
    private UUID preferencesId;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false, unique = true)
    private User user;

    // ── Travel Style & Interests ──────────────────────────────

    @Enumerated(EnumType.STRING)
    @Column(name = "travel_style", length = 30)
    private TravelStyle travelStyle;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "favorite_categories", columnDefinition = "jsonb")
    private List<String> favoriteCategories;

    @Enumerated(EnumType.STRING)
    @Column(name = "activity_level", length = 20)
    private ActivityLevel activityLevel;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "cuisine_preferences", columnDefinition = "jsonb")
    private List<String> cuisinePreferences;

    // ── Trip Preferences ──────────────────────────────────────

    @Enumerated(EnumType.STRING)
    @Column(name = "preferred_climate", length = 20)
    private PreferredClimate preferredClimate;

    @Enumerated(EnumType.STRING)
    @Column(name = "trip_pace", length = 20)
    private TripPace tripPace;

    @Enumerated(EnumType.STRING)
    @Column(name = "accommodation_type", length = 20)
    private AccommodationType accommodationType;

    @Enumerated(EnumType.STRING)
    @Column(name = "transport_preference", length = 30)
    private TransportPreference transportPreference;

    // ── Constraints & Accessibility ───────────────────────────

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "dietary_restrictions", columnDefinition = "jsonb")
    private List<String> dietaryRestrictions;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "accessibility_needs", columnDefinition = "jsonb")
    private List<String> accessibilityNeeds;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "avoid_categories", columnDefinition = "jsonb")
    private List<String> avoidCategories;

    // ── Budget Details ────────────────────────────────────────

    @Column(name = "daily_budget", precision = 10, scale = 2)
    private BigDecimal dailyBudget;

    @Column(name = "budget_currency", length = 3)
    private String budgetCurrency;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "splurge_categories", columnDefinition = "jsonb")
    private List<String> splurgeCategories;

    // ── Language & Communication ──────────────────────────────

    @Column(name = "preferred_language", length = 5)
    private String preferredLanguage;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "spoken_languages", columnDefinition = "jsonb")
    private List<String> spokenLanguages;

    // ── Timestamps ────────────────────────────────────────────

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    protected void onCreate() {
        Instant now = Instant.now();
        if (this.createdAt == null)
            this.createdAt = now;
        if (this.updatedAt == null)
            this.updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = Instant.now();
    }
}

File: mobile/lib/features/profile/data/models/user_preferences.dart
/// Domain model for user travel preferences (GET /users/me/preferences).
class UserPreferences {
  final String preferencesId;
  final String userId;
  final String? travelStyle;
  final List<String> favoriteCategories;
  final String? activityLevel;
  final List<String> cuisinePreferences;
  final String? preferredClimate;
  final String? tripPace;
  final String? accommodationType;
  final String? transportPreference;
  final List<String> dietaryRestrictions;
  final List<String> accessibilityNeeds;
  final List<String> avoidCategories;
  final num? dailyBudget;
  final String? budgetCurrency;
  final List<String> splurgeCategories;
  final String? preferredLanguage;
  final List<String> spokenLanguages;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const UserPreferences({
    required this.preferencesId,
    required this.userId,
    this.travelStyle,
    this.favoriteCategories = const [],
    this.activityLevel,
    this.cuisinePreferences = const [],
    this.preferredClimate,
    this.tripPace,
    this.accommodationType,
    this.transportPreference,
    this.dietaryRestrictions = const [],
    this.accessibilityNeeds = const [],
    this.avoidCategories = const [],
    this.dailyBudget,
    this.budgetCurrency,
    this.splurgeCategories = const [],
    this.preferredLanguage,
    this.spokenLanguages = const [],
    this.createdAt,
    this.updatedAt,
  });

  /// Safe defaults when opening Edit Preferences before preferences are loaded.
  static UserPreferences defaultForDraft({String userId = ''}) {
    return UserPreferences(
      preferencesId: '',
      userId: userId,
    );
  }

  UserPreferences copyWith({
    String? preferencesId,
    String? userId,
    String? travelStyle,
    List<String>? favoriteCategories,
    String? activityLevel,
    List<String>? cuisinePreferences,
    String? preferredClimate,
    String? tripPace,
    String? accommodationType,
    String? transportPreference,
    List<String>? dietaryRestrictions,
    List<String>? accessibilityNeeds,
    List<String>? avoidCategories,
    num? dailyBudget,
    String? budgetCurrency,
    List<String>? splurgeCategories,
    String? preferredLanguage,
    List<String>? spokenLanguages,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return UserPreferences(
      preferencesId: preferencesId ?? this.preferencesId,
      userId: userId ?? this.userId,
      travelStyle: travelStyle ?? this.travelStyle,
      favoriteCategories: favoriteCategories ?? this.favoriteCategories,
      activityLevel: activityLevel ?? this.activityLevel,
      cuisinePreferences: cuisinePreferences ?? this.cuisinePreferences,
      preferredClimate: preferredClimate ?? this.preferredClimate,
      tripPace: tripPace ?? this.tripPace,
      accommodationType: accommodationType ?? this.accommodationType,
      transportPreference: transportPreference ?? this.transportPreference,
      dietaryRestrictions: dietaryRestrictions ?? this.dietaryRestrictions,
      accessibilityNeeds: accessibilityNeeds ?? this.accessibilityNeeds,
      avoidCategories: avoidCategories ?? this.avoidCategories,
      dailyBudget: dailyBudget ?? this.dailyBudget,
      budgetCurrency: budgetCurrency ?? this.budgetCurrency,
      splurgeCategories: splurgeCategories ?? this.splurgeCategories,
      preferredLanguage: preferredLanguage ?? this.preferredLanguage,
      spokenLanguages: spokenLanguages ?? this.spokenLanguages,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
